### DISCUSSION TITLE
ActiveStorage support?

### DISCUSSION THREAD

**Discussion by mbajur** [jrochkind/attr_json#231] [Ideas]
Hey! After thinking a lot about it, i think that's not really possible to implement that without heavy hacking of ActiveStorage internals but maybe i miss something so i will give it a shot - do you think that supporting AS could be somehow possible in any way? That might be tricky as attachment relies on parent record existence but who knows :)

**Comment by jrochkind** [comment]
Hi, thanks for your note!

I do not use ActireStorage myself, so I'm not sure -- I'm not sure I understand in what ways this would be integrated with ActiveStorage -- like attaching a stored file to a nested model?  Hmm, I honestly have no idea!

I personally am considering this gem fairly feature-complete at the moment. But if someone else figures out a way to approach this, please do share!  I would be willing to review a PR, but it would have to be with fairly minimal/elegant/maintainable code additions here!

**Comment by mbajur** [comment]
Hey! Thanks for your reply @jrochkind and apologies for the lack of mine. I paused the development of thing i've been working on using attr_json for some time but now i'm back and i think i [almost] nailed it.

Basically, to do that, i decided to take a following path: I created a new `attr_json :image, :binary` on my [nested]model. I added a file_field in the form for this `image` attribute and now, on record save/creation, i'm iterating through all the `image` attributes submitted by user, i create a blob record with whatever user submitted in there and i reassign this newly created blob record's `global_id` to the `image` attribute instead. However, i encountered a one blocked I'm not able to take care of - this approach works fine when i'm creating a new nested model entry but not when I'm updating an existing one - because there is apparently no way to tell if the `image` field has been changed or not. Let me demonstrate that with a code sample:

```ruby
class ContentBlock < ApplicationRecord
  include AttrJson::NestedAttributes

  class Section
    include AttrJson::Model
    attr_json :image, :binary
  end

  attr_json :sections, Section.to_type, array: true
  attr_json_accepts_nested_attributes_for :sections, reject_if: :all_blank

  before_save :resolve_images

  private

  def resolve_images
    sections.each do |section|
      # Here is the problematic part: How to tell if `image` attribute has changed? 
      # There is no such method as `*_changed?` so i can't use the default rails way of doing that. 
      # On the other hand, I apparently don't have access to `ContentBlock` record from the `Section` 
      # callback so i can't access the underlying json in the DB. However, even if i did, it would feel a bit 
      # hackish but maybe that's the only way. Or maybe nothing i wrote above is true and i have access 
      # to all that but i just missed that in the code and/or in the docs :)
      next unless section.image_changed?
      next if section.image.blank?

      section.image = Spree::FileUpload.create!(file: section.image).to_global_id.to_s
    end
  end
end
```

More clarification in the comment above. Thanks for any clues! 🙏

**Reply by jrochkind** [reply]
Hi!

So attr_json does have some support for "dirty" tracking to see if an `attr_json` attribute has changed. 

https://github.com/jrochkind/attr_json?tab=readme-ov-file#activerecord-attributes-and-dirty-tracking

I'm not sure if you didn't notice this (the docs could prob use some examples, they assume familiarity with standard Rails dirty tracking that even I don't currently have!) -- or if you've tried it and it didn't work succesfully?

**Reply by mbajur** [reply]
Hey! yeah, i saw that but the problem is i have to call that on parent record instance. Given my `before_save` callback from example above is defined on `Section` instance (which is not a parent record but a nested collection element) - i don't have access to parent record (`ContentBlock`) from this callback method. I would imagine being able to call `self.record` or something like that to be able to do that but there is no such option :) Or I'm not aware of it's existence

**Reply by jrochkind** [reply]
Ah, I see what you mean. 

I'm not sure how to get that in there, because the model would need to be tied into the parent records lifecycle in a way it isn't right now. 

It might be easier if we refactor AttrJson::Model to be more fully based on ActiveModel::Attributes, or if Rails proceeds with some of the Attributes development we mention in #222 , perhaps you could do it purely based on built-into Rails stuff without attr_json at all!

For now, I'm not sure I see a great path, although definitely thoughts, suggestions, and PR's for consideration are always welcome!

**Reply by mbajur** [reply]
Thank you for the clarification. I guess i will proceed my images on the controller/form level then! :)

**Reply by jrochkind** [reply]
In some of my projects, I just don't put something in attr_json if it doesn't work well there. If you want to use ActiveStorage, it's worth considering redesigning your intended database schema so the ActiveStorage stuff can be in the ordinary columns it expects -- I am not very familiar with AS myself, but I know Rails stuff can sometimes be somewhat inflexible at first. 

Probably something you have already thought of, but just thought I'd mention, it is something I have done. A table that has a JSON column *plus* some ordinary columns, *plus* maybe a foreign key to a to-one associated table, which itself can have JSON as well as ordinary columns.
